### PR TITLE
Upload db checkpoints to cloud object stores

### PR DESCRIPTION
**Pull request by sadhansood**
## Description 

This PR allows us to upload db checkpoints (and later state snapshots) to cloud object stores like S3, GCS and Azure Blob Store. Goal is to have different full nodes upload checkpoints+snapshots to different cloud stores independently.

## Test Plan 

Unit tests with local fs for now. Working on S3 setup with PEs to test with cloud object stores.

### CODE REVIEW

File: crates/sui-core/src/db_checkpoint_handler.rs [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+            .await?;
+
+        let next_epoch = if let Some((last_epoch, path)) =
+            remote_checkpoints_by_epoch.last_key_value()
+        {
+            let success_marker = path.child(SUCCESS_MARKER);
+            let get_result = self.output_object_store.get(&success_marker).await;
+            match get_result {
+                Ok(_) => last_epoch + 1,
+                Err(Error::NotFound { .. }) => {
```
Comment (mystenmark) [inline crates/sui-core/src/db_checkpoint_handler.rs:99]: i'm confused by this - the comments below indicate that we should delete the local dir if UPLOAD_COMPLETE_MARKER has been written, but this deletes if SUCCESS_MARKER is not found?

Comment (mystenmark) [reply crates/sui-core/src/db_checkpoint_handler.rs:99]: ok, i get it now - but why do we need to delete the dir before continuing? shouldn't we assume we can resume where we left off?

Comment (sadhansood) [reply crates/sui-core/src/db_checkpoint_handler.rs:99]: My main worry was one of the files being copied is only partly copied and application crashed/stopped. In which case on resuming copy we will skip over that file. I am also not sure if the object store api for different cloud vendors will throw an error if a bucket key already exists when we resume copy and that is why found it safe to delete those files first. Overall found it simple to delete and copy as a first pass. Happy to incorporate any suggestions.

**Review by mystenmark — APPROVED** [review]
one more question and LGTM

File: crates/sui-core/src/db_checkpoint_handler.rs (hunk 1 above)
Comment (sadhansood) [reply crates/sui-core/src/db_checkpoint_handler.rs:99]: Some of the object store implementations use multi part upload where they split the underlying data into smaller chunks. So I was not sure, if a file upload is an atomic operation on the cloud store or not (so to be safe I assumed it is not)

**Pull request merged**